src/core/lwt.ml

```diff
@@ -22,1307 +22,2746 @@
+   dependent promises.
+
+   Some of these dependencies are explicit to Lwt, e.g. the callbacks registered
+   by [Lwt.bind]. Others are not visible to Lwt, because the user can always
+   register a callback using a function like [Lwt.on_success], and use that
+   callback to complete another initial promise. All the explicit dependencies
+   are created by Lwt's own sequential and concurrent composition functions
+   (so, [Lwt.bind], [Lwt.join], etc). Whether dependencies are explicit or not
+   is relevant only to cancelation.
+
+
+   5. Cancelation
+
+   As described above, ordinary promise completion procedes from an initial
+   promise, forward along callbacks through the dependency graph. Since it
+   starts from an initial promise, it can only be triggered using a resolver.
+
+   Cancelation is a sort of dual to ordinary completion. Instead of starting at
+   an initial promise/resolver, cancelation starts at *any* promise. It then
+   goes *backwards* through the explicit dependency graph, looking for
+   cancelable initial promises to cancel -- those that were created by
+   [Lwt.task]. After finding them, cancelation completes them normally with
+   [Failed Lwt.Canceled], causing an ordinary promise completion process.
+
+   To summarize, cancelation is a way to trigger an *ordinary* completion of
+   promises created with [Lwt.task], by first searching for them in the promise
+   dependency graph (which is assembled by [Lwt.bind], [Lwt.join], etc).
+
+   This backwards search is triggered only by [Lwt.cancel]. It is also possible
+   for the user to cancel a promise directly by failing it with [Lwt.Canceled],
+   but in all cases where the user can do so, the search would be redundant
+   anyway -- the user has only two ways of directly failing a promise with
+   [Lwt.Canceled] (or any exception, for that matter):
+
+   - The user can create an initial promise, then fail it through its resolver.
+     The search is redundant because it would find only the same initial promise
+     to cancel.
+   - The user can create a trivial promise by calling [Lwt.fail Lwt.Canceled].
+     The search is again redundant; in this case it would find nothing to
+     cancel.
+
+   Note that there is a quirk: only promises created by [Lwt.task] are
+   susceptible to being canceled by [Lwt.cancel], but the user can manually
+   cancel initial promises created by both [Lwt.task] and [Lwt.wait].
+
+   Due to [Lwt.cancel], promise cancelation, and therefore completion, can be
+   initiated by the user without access to a resolver. This is important for
+   reasoning about state changes in the implementation of Lwt, and is referenced
+   in some implementation detail comments.
+
+
+   6. No I/O
+
+   The Lwt core deliberately doesn't do I/O. The completion loop stops running
+   once no promises can be completed immediately. It has to be restarted later
+   by some external process.
```

Comment:
I've clarified that not every pending promise is tracked in a global queue (it doesn't have to be global, it's just anything outside the core). Only the promises that represent blocked I/O. The rest of the pending promises are just retained in memory through references inside callbacks registered on those I/O promises, and transitively.

@raphael-proust: Indeed, I mentioned `Lwt_main.run` in the paragraph below this one, so that the reader can visit the `Lwt_unix` instance of this outer I/O loop. However, that I/O loop is not yet well-documented. I plan to change that, and I think we can add a clearer reference once there is something to read there.